util/memoryusagealarm/memoryusagealarm.go

```diff
@@ -0,0 +1,370 @@
+	} else {
+		record.serverMemoryQuota, record.err = memory.MemTotal()
+		if record.err != nil {
+			logutil.BgLogger().Error("get system total memory fail", zap.Error(record.err))
+			return
+		}
+		record.isServerMemoryQuotaSet = false
+	}
+	record.lastCheckTime = time.Time{}
+	record.lastRecordMemUsed = uint64(float64(record.serverMemoryQuota) * record.memoryUsageAlarmRatio)
+	tidbLogDir, _ := filepath.Split(config.GetGlobalConfig().Log.File.Filename)
+	record.baseRecordDir = filepath.Join(tidbLogDir, "oom_record")
+	if record.err = disk.CheckAndCreateDir(record.baseRecordDir); record.err != nil {
+		return
+	}
+	// Read last records
+	recordDirs, err := os.ReadDir(record.baseRecordDir)
+	if err != nil {
+		record.err = err
+		return
+	}
+	for _, dir := range recordDirs {
+		name := filepath.Join(record.baseRecordDir, dir.Name())
+		if strings.Contains(dir.Name(), "record") {
+			record.lastRecordDirName = append(record.lastRecordDirName, name)
+		}
+	}
+	record.initialized = true
+}
+
+// If Performance.ServerMemoryQuota is set, use `ServerMemoryQuota * MemoryUsageAlarmRatio` to check oom risk.
+// If Performance.ServerMemoryQuota is not set, use `system total memory size * MemoryUsageAlarmRatio` to check oom risk.
+func (record *memoryUsageAlarm) alarm4ExcessiveMemUsage(sm util.SessionManager) {
+	if !record.initialized {
+		record.initMemoryUsageAlarmRecord()
+		if record.err != nil {
+			return
+		}
+	} else {
+		record.updateVariable()
+	}
+	if record.memoryUsageAlarmRatio <= 0.0 || record.memoryUsageAlarmRatio >= 1.0 {
+		return
+	}
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	runtime.ReadMemStats(instanceStats)
+	if record.isServerMemoryQuotaSet {
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, record.err = memory.MemUsed()
+		if record.err != nil {
+			logutil.BgLogger().Error("get system memory usage fail", zap.Error(record.err))
+			return
+		}
+	}
+
+	// TODO: Consider NextGC to record SQLs.
+	if record.needRecord(memoryUsage) {
+		record.lastCheckTime = time.Now()
+		record.lastRecordMemUsed = memoryUsage
+		record.doRecord(memoryUsage, instanceStats.HeapAlloc, sm)
+		record.tryRemoveNoNeedRecords()
+	}
+}
+
+func (record *memoryUsageAlarm) needRecord(memoryUsage uint64) bool {
+	// At least 60 seconds between two recordings that memory usage is less than threshold (default 70% system memory).
+	// If the memory is still exceeded, only records once.
+	// If the memory used ratio recorded this time is 0.1 higher than last time, we will force record this time.
+	if float64(memoryUsage) <= float64(record.serverMemoryQuota)*record.memoryUsageAlarmRatio {
+		return false
+	}
+
+	interval := time.Since(record.lastCheckTime)
+	memDiff := int64(memoryUsage) - int64(record.lastRecordMemUsed)
+	if interval > 5*time.Second {
+		logutil.BgLogger().Warn("Record Memory Information.", zap.String("record time", time.Now().String()), zap.String("last record time", record.lastCheckTime.String()))
```

Comment:
We do not need line 716 and line 180, it'll be hard to read if we have too many logs.
We can return the reason why we need to record, and print the reason in `doRecord`

Reply:
> We do not need line 716 and line 180, it'll be hard to read if we have too many logs. We can return the reason why we need to record, and print the reason in `doRecord`

done.